### ISSUE TITLE
DataStore - `@model(timestamp: null)` broken. `subscriptionError Connection failed: when omitting timestamp from model `

### ISSUE THREAD

**Issue by charlieforward9** [aws-amplify/amplify-js#11826]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

GraphQL API, DataStore

### Amplify Categories

storage

### Environment information

<details>

```
  System:
    OS: macOS 12.6.7
    CPU: (4) x64 Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
    Memory: 119.95 MB / 8.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 18.16.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.6.7 - ~/Documents/Github/Agroview2.0/node_modules/.bin/npm
  Browsers:
    Chrome: 115.0.5790.170
    Safari: 16.5.2
  npmPackages:
    @fortawesome/fontawesome-svg-core: ^6.4.0 => 6.4.0 
    @fortawesome/free-brands-svg-icons: ^6.4.0 => 6.4.0 
    @fortawesome/free-regular-svg-icons: ^6.4.0 => 6.4.0 
    @fortawesome/free-solid-svg-icons: ^6.4.0 => 6.4.0 
    @fortawesome/react-fontawesome: ^0.2.0 => 0.2.0 
    @reduxjs/toolkit: ^1.9.5 => 1.9.5 
    @reduxjs/toolkit-query:  1.0.0 
    @reduxjs/toolkit-query-react:  1.0.0 
    @turf/turf: ^6.5.0 => 6.5.0 
    @types/compression: ^1.7.2 => 1.7.2 
    @types/express: ^4.17.14 => 4.17.17 
    @types/leaflet: ^1.9.3 => 1.9.3 
    @types/node: ^20.2.5 => 20.2.5 
    @types/react: ^18.0.28 => 18.2.8 
    @types/react-dom: ^18.0.3 => 18.0.11 
    @types/react-slider: ^1.3.1 => 1.3.1 
    @types/turf: ^3.5.32 => 3.5.32 
    @typescript-eslint/eslint-plugin: ^5.59.9 => 5.59.9 
    @typescript-eslint/parser: ^5.62.0 => 5.62.0 
    @vitejs/plugin-react: ^4.0.4 => 4.0.4 
    @vitejs/plugin-react-swc: ^3.0.0 => 3.3.2 
    algoliasearch: ^4.19.1 => 4.19.1 
    amazon-cognito-identity-js: ^6.2.0 => 6.3.2 
    amazon-cognito-identity-js/internals:  undefined ()
    autoprefixer: ^10.4.14 => 10.4.14 
    aws-amplify: ^5.3.7 => 5.3.7 
    compression: ^1.7.4 => 1.7.4 
    concurrently: ^8.2.0 => 8.2.0 
    cross-env: ^7.0.3 => 7.0.3 
    css-loader: ^6.8.1 => 6.8.1 
    debounced-drag:  1.0.0 
    eslint: ^8.47.0 => 8.47.0 
    eslint-plugin-react-hooks: ^4.6.0 => 4.6.0 
    eslint-plugin-react-refresh: ^0.3.4 => 0.3.5 
    express: ^4.18.2 => 4.18.2 
    i: ^0.3.7 => 0.3.7 
    instantsearch.css: ^8.0.0 => 8.0.0 
    leaflet: ^1.9.4 => 1.9.4 
    leaflet-draw: ^1.0.4 => 1.0.4 
    npm: ^9.6.7 => 9.6.7 
    postcss: ^8.4.24 => 8.4.24 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    react-instantsearch-hooks-web: ^6.47.1 => 6.47.1 
    react-leaflet: ^4.2.1 => 4.2.1 
    react-leaflet-draw: ^0.20.4 => 0.20.4 
    react-redux: ^8.0.7 => 8.0.7 
    react-slider: ^2.0.5 => 2.0.5 
    react-tabs: ^6.0.1 => 6.0.1 
    react-use-is-online: ^1.3.0 => 1.3.0 
    recharts: ^2.7.3 => 2.7.3 
    redux: ^4.2.1 => 4.2.1 
    redux-todos-with-undo-example:  0.0.0 
    redux-undo: ^1.0.1 => 1.0.1 
    shp-write: github:Agriculture-Intelligence/shp-write => 0.3.2 
    sirv: ^2.0.2 => 2.0.2 
    style-loader: ^3.3.3 => 3.3.3 
    tailwindcss: ^3.3.2 => 3.3.2 
    ts-node: ^10.9.1 => 10.9.1 
    typescript: ^5.1.3 => 5.1.3 
    vite: ^4.0.3 => 4.3.9 
  npmGlobalPackages:
    @aws-amplify/cli: 10.2.1
    corepack: 0.17.0
    gh-pages: 3.2.3
    graphql-cli: 4.1.0
    n: 9.1.0
    nodemon: 2.0.13
    npm: 9.5.1
    stable: 0.1.8
    tls-test: 1.0.0
    ts-node: 10.9.1
```

</details>

### Describe the bug

> [This is by design. You can set timestamps: null to disable this per docs: https://docs.amplify.aws/cli/graphql-transformer/directives#definition](https://github.com/aws-amplify/amplify-cli/issues/4397#issuecomment-635541585)

but in the link, there is no mention of the timestamps. 

```typescript
type RawImage @model(timestamps: null) @auth(rules: [{ allow: public }]) {
```

this is what my model directive was updated to

### Expected behavior

The expected behavior is to map `@model(timestamp: null)` to the function of disabling the requirement of an input into the `createdAt: String, updatedAt: String` input into DataStore for content that does not require that information in our use case.

### Reproduction steps

1. change `@model()` to `@model(timestamps: null)`
2. check console

### Code Snippet

Full error object: 

```json
{"message": 
     "Connection failed: {\"errors\":[{\"message\":\"Validation error of type FieldUndefined: Field 'createdAt' in type 'RawImage' is undefined @ 'onCreateRawImage/createdAt'\"},{\"message\":\"Validation error of type FieldUndefined: Field 'updatedAt' in type 'RawImage' is undefined @ 'onCreateRawImage/updatedAt'\"}]}"
}    
```

**Comment by charlieforward9** [comment]
@cwomack I am looking for an update on this.

Our app currently uses a promise to indicate when Datastore is ready to be interacted with:

```ts
this.isReady = new Promise((resolve, reject) => {
      // eslint-disable-next-line prefer-const
      let timeoutId: number;
      DataStore.start();
      Hub.listen("datastore", (hubData) => {
        const { event, data } = hubData.payload;
        verbose && console.log(`DataStore event: ${event}`, data);
        if (event === "ready") {
          resolve();
          clearTimeout(timeoutId);
        }
      });

      // Reject promise after 10 seconds
      timeoutId = window.setTimeout(
        () => reject("Datastore not ready after 10 seconds"),
        10000
      );
    });
```

this helps avoid issues where datastore is not fully synced when we make our initial calls. The problem is that due to the failed connection error, Datastore never emits `ready`, so this promise rejects.

**Comment by cwomack** [comment]
@charlieforward9, thank you for opening this issue.  I'll get this reviewed internally and get you a response as soon as I can.

**Comment by svidgen** [comment]
The DataStore codebase currently expects timestamp fields to be present. This is different than the raw API category &mdash; the context under which [the timestamp disabling guidance](https://github.com/aws-amplify/amplify-cli/issues/4397#issuecomment-635541585) was given.

I'm not 100% sure off the cuff whether DataStore's dependency on those fields is a *hard* dependency. I don't believe we're using them as part of the sync protocol. But, I can tell you DataStore codebase generally assumes that the fields exists. (Though it does allow for them to be renamed.) And, I would *generally* recommend leaving them in place, even if they're not strictly necessary.

Are you running into problems by leaving those fields on the models? Is this a hard blocker for you?

**Comment by charlieforward9** [comment]
@svidgen Thanks for the details!

> Is this a hard blocker for you? 

Not a hard blocker, we storing from a lambda using Python and the timestamps are two unnecessary fields for the data we want to store. 

Could I ask for the reason behind the recommendation for not removing them? Their omission seems to be a performance/cost optimization that becomes more apparent at scale. Which is one of my top priorities in my role on the team.

**Comment by svidgen** [comment]
I just finished some testing on my end to confirm my suspicions, and I ended up disproving myself! Looking at the area of code I had in mind, DataStore does assume the fields exist, but it doesn't fail if they're omitted. And, in my testing so far, I actually don't see any problems during the sync.

To test, I created this simple model with removed timestamp fields and watched everything sync successfully. I sent a few creations through via the AppSync console, watched them come down the wire in the WS subscription, and saw them appear in my IndexedDB table.

```graphql
type NonTimestampedThing
    @model(timestamps: null)
    @auth(rules: [ { allow: owner } ])
{
    id: ID!
    name: String
}
```

So, I apologize for missing this initially, but looking at your error more closely, your app code is probably just out of date. This particular validation error means the client is asking for fields that don't exist in the *deployed* schema. This could happen if you initially created the model with those fields and then removed the fields &mdash; but didn't regenerate and re-build the models into the client app.

Try running `amplify codegen models` to make sure the generated client-side schema matches the deployed schema. If that fails, I'll ask @cwomack to take over here and help you either troubleshoot this or produce more detailed repro steps.

> Could you explain the reason behind the recommendation for not removing them?

That recommendation is based solely on the ability to make sense of data freshness within app code without depending on Amplify's internal fields. If you ever need the fields later, it's not only painful to add them back in, but the data you'd backfill would be 50% *lies!* 😅 

> Their omission seems to be a performance/cost optimization that becomes more apparent at scale.

Though it's not relevant to solving your problem, I'm very curious to know if you have measurements showing a significant difference here &mdash; primarily for my own edification. (I haven't tested this difference.)

**Comment by charlieforward9** [comment]
I did `amplify codegen models` and noticed my `schema.js` removed the config for the timestamp fields, so that seem to have worked. But now, I am getting a strange issue with the validation of one of my internal fields.

```
datastore.ts:582 Uncaught (in promise) Error: Field uploadID is required
    at datastore.ts:582:11
    at datastore.ts:916:7
    at Array.forEach (<anonymous>)
    at datastore.ts:913:28
    at produce (immerClass.ts:94:14)
    at Model.copyOf (datastore.ts:895:18)
    at traverseModel (util.ts:221:39)
    at IndexedDBAdapter.ts:345:28
    at step (tslib.es6.js:147:23)
    at Object.next (tslib.es6.js:128:53)
```

the only model that requires `uploadID` is not being used anywhere on our frontend at this moment, so I am unsure what is causing this. I followed up with `amplify push` and `amplify pull` just to make sure everything was synchronized. 


> I'm very curious to know if you have measurements showing a significant difference here

Unfortunately, I do not. It just makes sense to me that less data stored in a database and moving over a network result in preferred metrics. Our use case will have hundreds/thousand Image objects all tied to a specific Upload, so we will have the timestamp consideration covered, and try to keep the Image data as lean as possible. 

With that said, some performance/cost documentation would definitely be helpful to have. That would be a great discussion to be a part of to deepen my understanding of the tools I use.
